src/lib/tecs/TECS.cpp

```diff
@@ -41,241 +46,403 @@ using math::constrain;
-	_SPE_rate_setpoint = _hgt_rate_setpoint * CONSTANTS_ONE_G; // potential energy rate of change
-	_SKE_rate_setpoint = _tas_state * _TAS_rate_setpoint; // kinetic energy rate of change
+	} else if (_percent_undersped > FLT_EPSILON && flag.airspeed_enabled) {
+		pitch_speed_weight = 2.0f * _percent_undersped + (1.0f - _percent_undersped) * pitch_speed_weight;
 
-	// Calculate specific energies in units of (m**2/sec**2)
-	_SPE_estimate = _vert_pos_state * CONSTANTS_ONE_G; // potential energy
-	_SKE_estimate = 0.5f * _tas_state * _tas_state; // kinetic energy
+	} else if (!flag.airspeed_enabled) {
+		pitch_speed_weight = 0.0f;
 
-	// Calculate specific energy rates in units of (m**2/sec**3)
-	_SPE_rate = _vert_vel_state * CONSTANTS_ONE_G; // potential energy rate of change
-	_SKE_rate = _tas_state * _tas_rate_filtered;// kinetic energy rate of change
+	}
+
+	// don't allow any weight to be larger than one, as it has the same effect as reducing the control
+	// loop time constant and therefore can lead to a destabilization of that control loop
+	weight.spe_weighting = constrain(2.0f - pitch_speed_weight, 0.f, 1.f);
+	weight.ske_weighting = constrain(pitch_speed_weight, 0.f, 1.f);
+
+	return weight;
+}
+
+void TECSControl::_updatePitchSetpoint(float dt, const Input &input, const SpecificEnergy &se, Param &param,
+				       const Flag &flag)
+{
+	SpecificEnergyWeighting weight{_updateSpeedAltitudeWeights(param, flag)};
+	/*
+	 * The SKE_weighting variable controls how speed and altitude control are prioritised by the pitch demand calculation.
+	 * A weighting of 1 givea equal speed and altitude priority
+	 * A weighting of 0 gives 100% priority to altitude control and must be used when no airspeed measurement is available.
+	 * A weighting of 2 provides 100% priority to speed control and is used when:
+	 * a) an underspeed condition is detected.
+	 * b) during climbout where a minimum pitch angle has been set to ensure altitude is gained. If the airspeed
+	 * rises above the demanded value, the pitch angle demand is increased by the TECS controller to prevent the vehicle overspeeding.
+	 * The weighting can be adjusted between 0 and 2 depending on speed and altitude accuracy requirements.
+	*/
+	// Calculate the specific energy balance rate demand
+	float seb_rate_setpoint = se.spe.rate_setpoint * weight.spe_weighting - se.ske.rate_setpoint * weight.ske_weighting;
+
+	// Calculate the specific energy balance rate error
+	float seb_rate_error = (se.spe.rate_error * weight.spe_weighting) - (se.ske.rate_error * weight.ske_weighting);
+
+	_debug_output.energy_balance_rate_error = seb_rate_error;
+	_debug_output.energy_balance_rate_sp = seb_rate_setpoint;
+
+	if (param.integrator_gain_pitch > 0.0f) {
+		// Calculate pitch integrator input term
+		float pitch_integ_input = seb_rate_error * param.integrator_gain_pitch;
+
+		// Prevent the integrator changing in a direction that will increase pitch demand saturation
+		if (_pitch_setpoint > param.pitch_max) {
```

Comment:
@KonradRudin Can _pitch_setpoint really be larger than  param.pitch_max given that you are constraining it below and I assume there is no other function altering the state?